# Daily Coding 2017-02-16



## Async

### each

`col values`
```
[ '1c6398a0-8ba6-11e6-bd43-a710ab8a6164',
  '385bc900-90f7-11e6-876f-719554daeebf',
  '7c75b340-6d92-11e6-9f3b-212917332ccd' ]
```


```javascript
async.each(col, (item, cb) => {
  connection.query(QUERY.Channel.RegisterGroup, ['U', group_id, item], (err, result) => {
      cb(err, result);
  });
}, (err, result) => {
  callback(err, result);
  });
```

* `col` 값은 위의 선언된 배열의 값이다.
* `col` 에있는 배열의 값을 하나씩 꺼내와서 `item` 으로 넘긴다.
* `cb(err, result)`는 아래의 `(err, result) =>{...}` 함수로 값을 보낸다
* `callback(err, result)` 값에 에러가 하나라도 있으면 종료 callback이 호출 된다. 즉 에러를 만나면 바로 ㅃㅃ이다.

**이작업은 어제의 작업의 개선 작업이다.**

**원본 소스**
```javascript
for (let i = 0; i < len; i++) {
  connection.query(QUERY.Channel.RegisterGroup, ['U', group_id, channel_id[i]], (err, result) => {
      if(err){
          callback(err, null);
      }
  });
}
callback(null, null);
```
**개선점**
* `callback` 이 보장받지 못하는 문제 해결
* 만복문을 돌때마다 `if(err)` 검사를 진행하지 않고 `callback`으로 에러가발생했을 경우 즉시 종료
* `callback`으로 result을 돌려주지 못한점 개선


## AWS S3 Multiple Upload

**개선전**
```javascript
const len = files.length;
let i;
for (i = 0; i < len; i++) {
  params.Key = key + files[i].name;
  params.Body = require('fs').createReadStream(files[i].path);

  s3.upload(params, (err, result) => {
    if (err) {
      callback(err, null);
    }
  });
}
callback(null, null);
```

**개선후**

```javascript
async.each(files, (file, cb) => {
  
  params.Key = key + file.name;
  params.Body = require('fs').createReadStream(file.path);

  s3.upload(params, (err, result) => {
      cb(err, result);
  });
}, (err, result) => {
  callback(err, result);
});
```
**개선점**
* `callback` 이 보장받지 못하는 문제 해결
* 만복문을 돌때마다 `if(err)` 검사를 진행하지 않고 `callback`으로 에러가발생했을 경우 즉시 종료
* `callback`으로 result을 돌려주지 못한점 개선
